torchtune/modules/vision_transformer.py

```diff
@@ -0,0 +1,337 @@
+
+        # internal modules
+        self.conv = nn.Conv2d(
+            in_channels=in_channels,
+            out_channels=embed_dim,
+            kernel_size=(patch_size, patch_size),
+            stride=(patch_size, patch_size),
+            bias=False,
+        )
+
+        self.ln_post = LayerNorm(embed_dim)
+        self.ln_pre = LayerNorm(embed_dim)
+
+        self.cls_token_embedding = CLSEmbedding(embed_dim)
+
+    def get_image_tokens_per_tile(self):
+        return self.patches_per_tile + 1  # +1 for CLS token
+
+    def forward(
+        self, images: torch.Tensor, aspect_ratio: Optional[torch.Tensor] = None
+    ) -> Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
+        """
+        Args:
+            images (torch.Tensor): If the image was tile-cropped in advance, tensor with shape
+                (bsz, n_tiles, n_channels, tile_size, tile_size). Otherwise, tensor with shape
+                (bsz, n_channels, tile_size, tile_size).
+            aspect_ratio (Optional[torch.Tensor]): If the image was tile-cropped in advance,
+                Tensor with shape (bsz, 2), representing the aspect ratio of the image
+                before tile-cropping, e.g. (2,1). Otherwise, it should be None.
+
+        Returns:
+            Union[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]: The output tensor(s).
+                If indices_return_hidden is None, it returns the token embeddings as a tensor
+                of shape (bsz, n_tiles, n_tokens, embed_dim). Otherwise, it returns a
+                tuple of tensors: (x, hidden_out), where hidden_out is a stack of hidden layers
+                of shape (bsz, len(indices_return_hidden), n_tiles, n_tokens, embed_dim).
+
+        Examples:
+
+            >>> from torchtune.modules.transforms.vision_utils.tile_crop import tile_crop
+            >>> from torchtune.modules import VisionTransformer
+            >>>
+            >>> num_channels = 3
+            >>> image_size = (800,400)
+            >>> tile_size = 400
+            >>> patch_size = 40
+            >>> patch_grid_size = tile_size // patch_size
+            >>>
+            >>> # for details about preprocessing, please check
+            >>> # torchtune.models.clip._transforms.CLIPImageTransform
+            >>>
+            >>> # create a random image
+            >>> image = torch.rand(num_channels, image_size[0], image_size[1])
+            >>>
+            >>> # (num_tiles, nch, h, w) -> (2, 3, 400, 400)
+            >>> tile_cropped_image = tile_crop(image, tile_size)
+            >>> aspect_ratio = torch.tensor([2,1])
+            >>>
+            >>> # make it a batch of 1 image
+            >>> batch_image = tile_cropped_image.unsqueeze(0)
+            >>> batch_aspect_ratio = aspect_ratio.unsqueeze(0)
+            >>>
+            >>> # For a detailed example, please check
+            >>> # torchtune.models.clip._position_embeddings.clip_vision_encoder
+            >>> # model = VisionTransformer(
```

Comment:
i didnt want to construct the whole VisionTransformer in the example. Are we good with that?

Reply:
Yeah personally I think this should be fine